[FLINK-10995][runtime] Copy intermediate serialization results only once for broadcast mode

zhijiangW opened the pull request:
## What is the purpose of the change

*The behavior of current channel selector is either for one channel or all the channels for broadcast mode. In broadcast mode, the intermediate serialization results would be copied into every `BufferBuilder` requested for every sub partition,  so this would affect the performance seriously especially in large scale jobs.*

*We can copy to only one target `BufferBuilder` and the corresponding `BufferConsumer` would be shared by all the sub partitions to improve the performance. For mixed operations with broadcast and non-broadcast,  we should finish the previous `BufferBuilder` first before transforming from broadcast to non-broadcast, vice versa.*

## Brief change log

  - *Make `RecordWriter` as abstract class to extract some abstract methods*

  - *Implement `ChannelSelectorRecordWriter` and `BroadcastRecordWriter` based on new abstractions*

  - *Refactor `BufferBuilder` to support creating multiple `BufferConsumer`s*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added new `BroadcastRecordWriterTest` for verifying the mixed operations*

  - *Modify some previous unit tests*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zhijiangW commented:
@pnowojski , both `RecordWriter` and `BroadcastRecordWriter` are involved in broadcast operations. In order to improve the serialization copy process for all kinds of broadcast, I make the changes on `RecordWriter` side instead of special path only on `BroadcastRecordWriter`.  So it brings two different paths in some logics on `RecordWriter`.

In addition, the failed travis seems un-related with this PR change.

pnowojski commented:
@zhijiangW in what scenarios is the `RecordWriter` being used for broadcasting messages? I'm asking since it might not be worth paying the price of more complex code to support this.

zhijiangW commented:
@pnowojski , thanks for the kindly suggestion!

The current broadcast usages for `RecordWriter` is only for emitting `Watermark` and `StreamStatus`.  Your worries might be right and I also ever considered that it might bring bad effects for special scenarios if switching frequently.

Maybe the conservative way is only making the special improvement path for `BroadcastRecordWriter` which switches the mode in emitting `LatencyMarker` currently. If you approve this way, I would adjust the codes then. :)

pnowojski commented:
`Watermark` and `StreamStatus` I was aware of, I was a bit unsure that you have found some other use cases :)

I wouldn't expect that emitting `Watermark` or `StreamStatus` can be a performance bottleneck, except of some super skewed/weird use cases, so I don't think that both complicating the code and possibly worsening performance in other cases is worth it. So I would stick with optimizing just the `BroadcastRecordWriter`.

zhijiangW commented:
@pnowojski , thanks for your review and suggestions!  I have updated the codes for addressing the comments.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterTest.java:
```diff
@@ -247,136 +388,121 @@ public void testBroadcastEventMixedRecords() throws Exception {
 		writer.broadcastEvent(barrier);
 
-		assertEquals(4, bufferProvider.getNumberOfCreatedBuffers());
+		if (isBroadcastWriter) {
```
with `BroadcastRecordWriterTest` class, this if could be changed to overloaded method.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterTest.java:
```diff
@@ -247,136 +388,121 @@ public void testBroadcastEventMixedRecords() throws Exception {
+			RecordWriter.createRecordWriter(partition);
 
-		writer.broadcastEvent(EndOfPartitionEvent.INSTANCE);
+		if (isBroadcastEvent) {
```
+1 for the code deduplication, but could you do it in separate commit (preferably before the actual change)? Now it's confusing which changes are functional and which are caused by deduplication/refactoring.

zhijiangW replied:
Yes, it would confuse you after I refactor the whole tests.
I could keep the previous tests unchanged, but if my new tests reuse some common codes from previous tests. Do you mind I extract the common parts together?  For the old tests not related to my new tests, I would keep them unchanged.

pnowojski replied:
I'm not saying that you should revert this code deduplication (definitely keep it!), but just extract it to a separate commit, to not mix refactor and functional changes.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterTest.java:
```diff
@@ -461,6 +572,56 @@ private void emitRecordWithBroadcastPartitionerOrBroadcastEmitRecord(boolean isB
 		}
 	}
 
+	private void randomEmitMixedBroadcastBufferOrEvent(boolean isBroadcastEvent) throws Exception {
```
I'm not sure if I understand the idea behind the unit tests in this file. What's the point of asserting the count and the size of buffers? In doesn't check for the correctness, nor it checks for the performance it just checks for some private implementation detail.

For performance testing we have micro benchmarks.

For correctness we should only check that certain scenarios like, writing x, y, z records or events and/or mixed with broadcasting records/events/watermarks, result with the correct output content of the buffers, regardless of the number/size of the buffers on the output. Otherwise if we optimise/refactor something, all of those tests will start yielding false positive errors.

I'm not sure if you understand my concern/agree with it. If yes, and If you do not want to fix the old tests, could you please at least re-write the new test? This `randomEmitMixedBroadcastBufferOrEvent` test shouldn't care for number of `getNumberOfCreatedBuffers`.

zhijiangW replied:
My motivation of adding this test is for verifying the logic after switching modes between broadcast and random. That is we have to finish the current `BufferBuilder` once switching modes, and this logic can be reflected on the number of buffers in all the partition queues.

I could understand your concern of  best-effort ignoring the previous tests failure after refactoring something, and I agree with this rule. If we change this new test into data correctness like we already have in `RecordWriterTest#emitOrBroadcastRecord`, it might be difficult to verify how many `LatencyMarker` are emitted actually because this marker would not affect data correctness sometimes. I mean comparing `broadcastEmit` maker with `randomEmit` marker, the results might be same in aspect of data correctness, but the behavior is different.  I would have a try to re-write this test by both verifying data correctness and the number of markers in channel if you think this way make sense.

pnowojski replied:
Yes, we can verify the number of sent/received `LatencyMarker`s, I'm fine with that, but I would expect it to be done via deserilizing and not just based on the "number of buffers".

zhijiangW replied:
Agree

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/BroadcastRecordWriter.java:
```diff
@@ -44,4 +52,67 @@ public BroadcastRecordWriter(
 	}
+
+	@Override
+	public void broadcastEmit(T record) throws IOException, InterruptedException {
```
I see couple of issues/potential issues here:
1. I'm not sure if I like the `randomTriggered` state field and how it's value affects the logic in different methods. It took quite some time to understand two successive `randomEmit` one after another work correctly (and how they actually work), because `requestNewBufferBuilder` behaves differently for `randomTriggered = true`. 
2. I'm not sure, but I think that `BroadcastRecordWriter`'s logic would be actually easier to follow, if it didn't have `bufferBuilders` array, but single `bufferBuilder` field. Maybe it would be worth creating `AbstractRecordWriter` without `bufferBuilders`/`bufferBuilder` field and then have two concrete implementations of `RecordWriter` and `BroadcastRecordWriter` with one two different implementations?
3. `LatencyMarker`s are a bit broken here, since effectively they will be flushed on the very first next emit following the `LatencyMarker`, so they do not measure the actual latency. This might not be the critical issue, but it would be good to have correct metric for that.

Point 1. is not that difficult to address (for 1. `randomEmit()` could flush the current buffer, request a new one for one `randomEmit()` and immediately finish it, so that this special handling of `randomTriggered = true` wouldn't leak outside of the `randomEmit()` method).

Point 2. Might be not worth doing so.

Point 3. is the more problematic one I guess. 

a) One way to deal with it, would be to broadcast `LatencyMarker`. Easy to do, but with huge number of output channels and multiple broadcasts channels one after another, would result with gigantic cascade/avalanche of the `LatencyMarkers`. 
b) Another idea would be to broadcast the `LatencyMarker` to everybody, but add some kind of marker to let the receiver know, that it should ignore it, because it was targeted to someone else.
c) Maybe there are other solutions that I haven't thought about?

zhijiangW replied:
Thanks for these kindly detail suggestions. You pointed out a very critical potential problem which I have not thought through before.

Regarding with the `LatencyMarker`, I have not thought of a perfect solution yet.
1. If finishing the `LatencyMarker` immediately, it could only reflect the latency before this marker, and lose a bit latency of current `BufferConsumer` containing `LatencyMarker`.

2. If broadcasting the `LatencyMarker`, there are two concerns: First we might add the `LatencyMarkerID` in the marker like `CheckpointID`, and ignore the duplicate id in `StreamInputProcessor` which needs maintain a set of `LatencyMarkerID` for filter.  We could add this logic in processor to avoid gigantic avalanche of the marker. If to do so, the logic seems very easy for `BroadcastRecordWriter`, and the methods for `emit` and `randomEmit` are all processed in broadcast mode. But if we extend other elements to be `randomEmit` future, we might need add corresponding special handing logic like `LatencyMarker` then. The second concern is that broadcasting `LatencyMarker` still loses the latency precision compared with now. Currently we choose the random channel to send this marker, so it could reflect the average latency from statistics. But the broadcast mode could always reflect the latency of the fastest channel, because once the downstream receives the first marker, it would ignore the same following markers from other slow channels.

Considering the above two options, I would prefer the first option because the latency lose seems less compared with broadcasting way, and we do not need to adjust the logic on receiver side (how many IDs to be maintained in memory might be another concern in this way). 

Regarding with the first option, I agree with your suggestion of abstracting `AbstractRecordWriter` to maintain single `BufferBuilder` in `BroadcastRecordWriter`.  And  I could understand your concerns of the `randomTriggered` field. Your suggestion of finishing `LatencyMarker` immediately in `randomEmit` could avoid this field. I would like to have a try based on them if you also approve the first option. :)

zhijiangW replied:
Thought of another two points:
- `randomTriggered` field could not be removed completely, because the behaviors are different in the methods of `requestNewBufferBuilder` and `flushTargetPartition` based on this tag. Otherwise the current `copyFromSerializerToTargetChannel` could not be reused and `BroadcastRecordWriter` has to implement the private process of `randomEmit`. If we could retain this tag field, and finish current `BufferBuilder` immediately after `randomEmit`. Then only `randomEmit` would change this field, and only `requestNewBufferBuilder` and `flushTargetPartition` would read this tag. It would seem easy to understand compared with now.

- Considering immediately finishing current `BufferBuilder` once `randomEmit` done, it might increase the frequency of finishing `BufferBuilder`. For example, if the `randomEmit` is triggered very frequently and there are no common `emit` or `broadcastEmit` calling for a while, one `BufferBuilder` might write multiple random emitted elements, no need one `BufferBuilder` for one `randomEmit`. Or this worry is not worth considering for most scenarios?

pnowojski replied:
I also do not like spilling this logic to the receiver. I thought a little bit more about it and there is also one more option that should work. 

We have one shared `BufferBuilder` with multiple `BufferConsumers`. Once we get `randomEmit()` for a single channel `channelId`, we:
1. finish the current `BufferBuilder`
2. we create a new `BufferBuilder` (let's call it `bb`)
3. we serialize `LatencyMarker` to the `bb`
4. we enqueue new `BufferConsumer` created from `bb` to the `channelId`
5. For the sake of other channels, we create `BufferConsumer`s from the `bb` with offsetted starting position. Starting position should be offseted in a such way, that the `LatencyMarker` wouldn't be visible for those channels. (initiating `BufferConsumer#currentReaderPosition` to `LatencyMarker.size()`).

There would be one drawback of this solution. Reported latency, would be slightly higher, since latency markers would be always written to the buffer's beginning, thus marker would have to wait a bit more for the buffer to fill in compared to how long the regular records do wait on average. However flushing timeout wouldn't be affected, so in the end this should be fine.

What do you think?

zhijiangW replied:
Thanks for thinking of one more possible solution for this issue. I actually like the idea of sharing random `BufferBuilder` in all partitions by setting the offset position of `BufferConsumer`. In the before I also thought of another similar way of always sharing `BufferBuilder` no matter with broadcast switch, then every `BufferConsumer` maintains the different skip list to filter the data not belonging to itself. But it seems so complicated and might bring extra overheads. Your suggestion make it easy to only skip fixed length at the beginning of `BufferConsumer`.

I am not so worried about the latency higher as you mentioned because we also have flushing mechanism to work around. But I am just confused of another issue, that means we strongly couple  the `randomEmit` with `LatencyMarker` together. The`randomEmit` should be a common method provided for operators, but the implementation seems not extendable,  even though it might never have other usages for `randomEmit` in future.

zhijiangW replied:
Regarding with abstracting `RecordWriter`, I have some plans for your confirm.

1. Make current `RecordWriter` as abstract class, so it is no need to refactor existing references in other places. Then it has two different implementations `BroadcastRecordWriter` and `SelectorRecordWriter` separately.

2.  There are two main differences for these two implementations. `BroadcastRecordWriter` maintain single `BufferBuilder` and no need `ChannelSelector`. `SelectorRecordWriter` maintains the array of `BufferBuilder` and has the `ChannelSelector`.

3. Based on above two differences, we might need to abstract related methods such as `emit`, `randomEmit`, `broadcastEmit`, `getBufferBuilder`, `tryFinishCurrentBufferBuilder`, `requestNewBufferBuilder`, `closeBufferBuilder`.

4. In order to reuse the common main process of `RecordWriter#copyFromSerializerToTargetChannel` for all kinds of emit, we still need the boolean flag `randomTriggered` in `BroadcastRecordWriter`, because it would have two different paths in `requestNewBufferBuilder` and `flushTargetPartition` based on this condition. And this flag would only be updated to true at the beginning of `BroadcastRecordWriter#randomEmit`, and updated to false at the end of this method.

Do you have other concerns for above thoughts ?

pnowojski replied:
I don't think that we strongly couple `randomEmit` and `LatencyMarker` with my last proposal. The only "bad" thing that I see which we would be doing is that `randomEmit()` will have a performance penalties if used on `BroadcastRecordWriter`, which I guess is an acceptable trade off. If we start using `randomEmit()` for more use cases, we might just need to make selection of regular vs `BroadcastRecordWriter` either smarter or configurable.

points 1 to 3 make sense to me. I don't fully understand the issue with `copyFromSerializerToTargetChannel`, but this is a small change, so it might be just better if you write the code and show it in the PR instead of explaining.

zhijiangW replied:
Considering initiating `BufferConsumer#currentReaderPosition` to `LatencyMarker.size()` for other channels in `BroadcastRecordWriter#randomEmit`, I meant this way of getting marker size seems coupled with `LatencyMarker` currently. Maybe we do not need consider it now, if there are other usages in future, we can make it smarter then.

For `copyFromSerializerToTargetChannel`, this process might be called either by `randomEmit` or `BroadcastRecordWriter#emit`. And the `requestNewBufferBuilder` would be triggered in `copyFromSerializerToTargetChannel`. If it is in the `randomEmit` process, we need to initialize `BufferConsumer#currentReaderPosition` for other sharing channels, otherwise we do not need to do. So it needs a flag to distinguish the different ways.  It would be easy for explanation after I submit the codes later. :)

pnowojski replied:
> Considering initiating BufferConsumer#currentReaderPosition to LatencyMarker.size() for other channels in BroadcastRecordWriter#randomEmit, I meant this way of getting marker size seems coupled with LatencyMarker currently. 

We don't have to hardcode `LatencyMarker.size()`. After all when we were about to set the offset, we have just serialized the record for `randomEmit()` to a fresh `BufferBuilder` :)

zhijiangW replied:
Yes, you are right. We do not need hardcode the size of latency marker. :)

I have submitted the codes in a separate commit. I am not sure whether this changes is suitable based on our comments, so I only refactored the main processes and ignored the tests with annotations. After your review and confirm, I would refactor the commits and tests.

I think it is better to explain some thoughts below:

In order to reuse the main process of `RecordWriter#copyFromSerializerToTargetChannel`, the `randomTriggered` field is still maintained in `BroadcastRecordWriter`, but I think the logic is easy to understand than the first version.

After abstracting the `RecordWriter` to separate two implementations, the changes seem more complex, because there are many abstract methods involved in `BufferBuilder` and `ChannelSelector` which are the main differences between two implementations. 

I am not sure you could accept the way of magic 0 in `BroadcastRecordWriter#broadcastEmit` for reusing the common codes. If we do not pass the target channel in broadcast operation, we might need to realize all the related processes in `BroadcastRecordWriter`.

I think my first version seems more easy to follow if we only change the process of `BroadcastRecordWriter#randomEmit` as now. Maybe you have other good ways. :)

pnowojski replied:
Ok, I see the problem now with `randomTriggered`. I don't like it, but also I don't have anything better to suggest :( So let it be.

> After abstracting the RecordWriter to separate two implementations, the changes seem more complex

true. The alternate solution would be just dump the base class, but I don't think that it would be better. 

> I am not sure you could accept the way of magic 0

I noticed it before and again I'm not a big fan of it, but I can not find a better solution.

> think my first version seems more easy to follow

I'm not sure. It had less code/changes. Issue is that in that previous version, `BroadcastRecordWriter` was still using array of `BufferBuilder`s and this is also confusing and probably less performant. If you could run network benchmarks showing that both versions are equally performant I would be ok-ish with either one.

I was expecting this to be a simple couple of lines change, but clearly I have missed judged it :(

zhijiangW replied:
The fist version is just trying best to reuse the common codes. But it seems not good to understand and distinguish these two writers in essential, including `BufferBuilder` and `ChannelSelector`. Furthermore in theory the single `BufferBuilder` seems more efficient for `BroadcastRecordWriter`. From this point, I am willing to keep the current version if there are no better solutions currently. 

I would refactor the commit and related tests based on current idea if you have no other concerns. :)

pnowojski replied:
Sounds good :)

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterTest.java:
```diff
@@ -210,10 +354,12 @@ public void testBroadcastEventNoRecords() throws Exception {
+			RecordWriter.createRecordWriter(partitionWriter, new OutputEmitter(ShipStrategyType.BROADCAST, 0), "test") :
+			RecordWriter.createRecordWriter(partitionWriter);
+
+		CheckpointBarrier barrier = new CheckpointBarrier(0, 0, CheckpointOptions.forCheckpointWithDefaultLocation());
```
Just for simplify the codes. :) 
This line is so long because of the `CheckpointID` and timeout info which seem not helpful for the test function, so I refer to other tests to change them to 0 simple.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterTest.java:
```diff
@@ -196,10 +198,152 @@ public void testSerializerClearedAfterClearBuffers() throws Exception {
 	 */
 	@Test
-	public void testBroadcastEventNoRecords() throws Exception {
+	public void testBroadcastWriterBroadcastEventNoRecord() throws Exception {
```
I considered this way before and also preferred it actually.

But I wonder maybe not all the existing tests in `RecordWriterTest` are necessary for executing  in `BroadcastRecordWriter`, like `testClearBuffersAfterInterruptDuringBlockingBufferRequest` and `testSerializerClearedAfterClearBuffers` which I have not touched before. Then I only added the broadcast modes for the other existing tests.

I would double check and try to refactor the tests in a separate `BroadcastRecordWriterTest`.

pnowojski replied:
You can also override a base test with an empty method, marked overridden method `@Ignore` or have abstract base class, with two concrete `RecordWriterTest` and `BroadcastRecordWriterTest`. But maybe you are right, that the approach that you've taken now might be simpler/easier, so you would have to play with the code and use your own judgement here.

zhijiangW replied:
Thanks for the suggestion! I would have a try and it could be better if all the existing tests are suitable for both `BroadcastRecordWriter` and `RecordWriter. :)

zhijiangW commented:
@pnowojski , I updated the codes addressing above comments. :)

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java:
```diff
@@ -40,26 +40,33 @@
+			memorySegment,
+			recycler,
+			positionMarker,
+			positionMarker.position);
```
I think you could use `positionMarker.cachedPosition` here. `position` is marked `volatile` for the `BufferConsumer`'s thread. Here we are in the realm of `BufferBuilder`'s thread.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java:
```diff
@@ -40,26 +40,33 @@
+	 * @return created matching instance of {@link BufferConsumer} similar with {@link #createBufferConsumer()},
+	 * except for initializing its reader position based on {@link BufferBuilder}'s current writable position .
+	 */
+	public BufferConsumer createPositionedBufferConsumer() {
```
I guess you could drop this method and change the contract of `createBufferConsumer` to:
```
/**
  * (...)
  * This method always creates a `BufferConsumer` starting from the current writer offset. Data written to
  * BufferBuilder before creation of BufferConsumer won't be visible for that BufferConsumer
  **/
public BufferConsumer createBufferConsumer() {
		return new BufferConsumer(
			memorySegment,
			recycler,
			positionMarker,
			positionMarker.cachedPosition);
}
```

and this needs to be covered in the `BufferBuilderAndConsumerTest`

zhijiangW replied:
Yes, I would supplement a unit test for it.

zhijiangW replied:
I found if we unify the `createBufferConsumer` as above, it would affect the previous tests which creates `BufferConsumer` after the `BufferBuilder` is filled. I need to refactor these tests to confirm create `BufferConsumer` before writing.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/BroadcastRecordWriterTest.java:
```diff
@@ -21,30 +21,30 @@
 import org.apache.flink.core.io.IOReadableWritable;
```
Is this commit fixing tests broken by the previous commit? Or is it refactor? If the first this commit be squashed with the previous one. If the latter could you add an better explanation what this commit is doing in the commit message?

zhijiangW replied:
I guess you meant the third commit in this PR? I should explain the differences for these commits.

The second commit adds one whole test for `BroadcastRecordWriter`.

The third commit is just making the previous tests in `RecordWriterTest` work in the way of `BroadcastRecordWriter`. So it is a refactor work. Before refactor, all the tests would create `SelectorRecordWriter` instance when running in `BroadcastRecordWriterTest` . After refactor, we defined the `isBroadcastWriter` for creating `BroadcastRecordWriter` instance when running these tests in `BroadcastRecordWriterTest`. It could be squashed with the second commit. But I thought it would be clear for review separately.

The fourth commit is for simplifying the previous codes, actually it is not related to this PR. We could also submit a separate PR for it.

zhijiangW replied:
I think I would squash the second with third commit. And submit a hotfix PR separately for the fourth commit. What do you think. The first commit should be solved in previous PR.

pnowojski replied:
> And submit a hotfix PR separately for the fourth commit. 

I'm fine with having hotfixes in one single PR

> The third commit is just making the previous tests in RecordWriterTest work in the way of BroadcastRecordWriter

So the `BroadcastRecordWriterTest` added in the 2nd commit, doesn't actually test the `BroadcastRecordWriter` until it's fixed in the 3rd commit? In that case definitely those two commits should be squashed.

zhijiangW replied:
Not really, the new added test in `BroadcastRecordWriterTest` for 2nd commit actually tests the function for `BroadcastRecordWriter`. In this commit, the `BroadcastRecordWriterTest` also extends the `RecordWriterTest`, so it would run all the existing tests in `RecordWriterTest`, but the created writer is still the `SelectorRecordWriter`. 

In short words, `BroadcastRecordWriterTest` extends all the existing tests in `RecordWriterTest` and run them in the way of `SelectorRecordWriter`. So in the 3rd commit, I introduced the `isBroadcastWriter` in the constructor for creating the specific `RecordWriter` instance based on this tag. Then all the existing tests would run for `BroadcastRecordWriterTest` in the way of `BroadcastRecordWriter`.

Anyway, I think it still makes sense to squash these two commits. The separation is just for distinguishing the new added test and refactoring all the old tests.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterTest.java:
```diff
@@ -311,7 +334,7 @@ public void testBroadcastEventBufferReferenceCounting() throws Exception {
 		ResultPartitionWriter partition =
 			new CollectingPartitionWriter(queues, new TestPooledBufferProvider(Integer.MAX_VALUE));
-		RecordWriter<?> writer = RecordWriterBuilder.createRecordWriter(partition);
+		RecordWriter<?> writer = createRecordWriter(partition);
```
If this is a test fix, not a refactor, this change should be in separate commit. However I think this is going to be changed anyway, because of my pending  "Builder pattern" comment/issue.

zhijiangW commented:
@pnowojski , thanks for review again! 
I rebase the latest codes and squash the commits for addressing above comments.

pnowojski commented:
@zhijiangW do we want to revisit this one? I came across this, when I was looking for this code in the master branch (I was sure that it has been merged) while checking out things for:
> [DISCUSS] Enhance Support for Multicast Communication Pattern

zhijiangW commented:
Yes, we could revisit this PR soon. :)

zhijiangW commented:
@flinkbot run travis

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/BroadcastRecordWriter.java:
```diff
@@ -18,30 +18,163 @@
 public class BroadcastRecordWriter<T extends IOReadableWritable> extends RecordWriter<T> {
 
+	/** The current buffer builder shared for all the channels. */
+	private Optional<BufferBuilder> bufferBuilder = Optional.empty();
```
Optional fields are [discouraged](http://mail-archives.apache.org/mod_mbox/flink-dev/201908.mbox/%3CCAA_61Xo9oif8RjJgvxFNo%2Bua7_DoFksCy_5c_NOnQLWfS4-8qA%40mail.gmail.com%3E).

zhijiangW replied:
I might be out of date for the recent suggestions of code style. Actually this code was written one year ago and I do not adjust it based on new rules now. I would refer to the new code style and discard the `Optional` way if not suggested. :)

pnowojski replied:
I think the decision was to enforce code style changes only for new PRs, not the pending ones, so I wouldn't enforce the change here.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ChannelSelectorRecordWriter.java:
```diff
@@ -0,0 +1,142 @@
+
+	private final ChannelSelector<T> channelSelector;
+
+	private final Optional<BufferBuilder>[] bufferBuilders;
```
See above, no optional fields. Also mixing generics and arrays is not recommended.

zhijiangW replied:
Yes

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/RecordWriter.java:
```diff
@@ -54,23 +54,17 @@
+	protected final ResultPartitionWriter targetPartition;
 
-	private final ChannelSelector<T> channelSelector;
+	protected final int numberOfChannels;
```
any reason to switch the order of fields here?

zhijiangW replied:
It was not my intention for switching the order here. I only changed the `numberOfChannels` as `protected` and added one more field `RecordSerializer<T> serializer` below.

I guess if I added the `serializer` field in other regions it would avoid this issue. I would have a try.

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/BroadcastRecordWriter.java:
```diff
@@ -18,30 +18,163 @@
  *
  * @param <T> the type of the record that can be emitted with this record writer
  */
 public class BroadcastRecordWriter<T extends IOReadableWritable> extends RecordWriter<T> {
```
final for JIT optimization?

zhijiangW replied:
Not quite sure here, because I saw most of the existing cases were not defining the final for class case.

pnowojski replied:
(final has no effect on optimizations by itself)

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ChannelSelectorRecordWriter.java:
```diff
@@ -0,0 +1,142 @@
+ *
+ * @param <T> the type of the record that can be emitted with this record writer
+ */
+public class ChannelSelectorRecordWriter<T extends IOReadableWritable> extends RecordWriter<T> {
```
final for JIT optimization?

AHeise commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/RecordWriter.java:
```diff
@@ -225,46 +178,56 @@ public void setMetricGroup(TaskIOMetricGroup metrics) {
+	abstract void emptyCurrentBufferBuilder(int targetChannel);
+
+	/**
+	 * Marks the current {@link BufferBuilder} as finished and empty for the target channel.
```
suggestion: Marks the current {@link BufferBuilder} as finished and releases the resources for the target channel.

AHeise commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/BroadcastRecordWriterTest.java:
```diff
@@ -0,0 +1,98 @@
+	 * {@link BroadcastRecordWriter#randomEmit(IOReadableWritable)}.
+	 */
+	@Test
+	public void testBroadcastMixedRandomEmitRecord() throws Exception {
```
I couldn't understand the test after trying 3 times. Could you please add some comments in between?

zhijiangW replied:
Sure, I would do that.

zhijiangW commented:
Thanks for the review @AHeise and I would consider fixing the `optional` issue as you suggested.

zhijiangW commented:
Thanks for the reviews @pnowojski @AHeise ! I updated the codes for addressing comments. 

For the issue of `Optional` I would like to adjust it in a future separate ticket also including the previous `outputFlusher`, since both of you agree that.

zhijiangW commented:
@flinkbot run travis
